src/db_impl_gc.cc

```diff
@@ -202,21 +202,26 @@ Status TitanDBImpl::BackgroundGC(LogBuffer* log_buffer,
     ROCKS_LOG_BUFFER(log_buffer, "Titan GC nothing to do");
   } else {
     StopWatch gc_sw(env_, statistics(stats_.get()), TITAN_GC_MICROS);
-    BlobGCJob blob_gc_job(blob_gc.get(), db_, &mutex_, db_options_,
-                          gc_merge_rewrite, env_, env_options_,
-                          blob_manager_.get(), blob_file_set_.get(), log_buffer,
-                          &shuting_down_, stats_.get());
-    s = blob_gc_job.Prepare();
+    BlobGCJob *blob_gc_job = new BlobGCJob(blob_gc.get(), db_, &mutex_, db_options_,
+                                           gc_merge_rewrite, env_, env_options_,
+                                           blob_manager_.get(), blob_file_set_.get(), log_buffer,
+                                           &shuting_down_, stats_.get());
+    s = blob_gc_job->Prepare();
     if (s.ok()) {
       mutex_.Unlock();
       TEST_SYNC_POINT("TitanDBImpl::BackgroundGC::BeforeRunGCJob");
-      s = blob_gc_job.Run();
+      s = blob_gc_job->Run();
       TEST_SYNC_POINT("TitanDBImpl::BackgroundGC::AfterRunGCJob");
       mutex_.Lock();
     }
     if (s.ok()) {
-      s = blob_gc_job.Finish();
+      s = blob_gc_job->Finish();
     }
+
+    mutex_.Unlock();
+    delete blob_gc_job;
```

Comment:
not sure if unlocking would cause any race condition, and it is just hard to check. A better pattern is we can pass a `struct GCContext { std::unique_ptr<BlobGCJob> }` from outside of the mutex (in `BackgroundCallGC()`) into `BackgroundGC` and use the context struct to hold the pointer, and make sure the context struct is cleanup outside of mutex.

Reply:
We checked the BlobGCJob destruct function and sure it will not cause race condition, but it is a good idea to use GCContext to manage the BlobGCJob pointer.

Reply:
Not just `BlobGCJob` destructor. I didn't check whether it is safe if some other job slip in between `blob_gc_job->Finish()` and `blob_gc->ReleaseGcFiles()`.